torch/csrc/jit/script/compiler.cpp

```diff
@@ -1272,18 +1272,44 @@ struct to_ir {
     }
   }
 
+  bool isIsInstanceCall(const Expr& expr) {
+    if (expr.kind() != TK_APPLY) {
+      return false;
+    }
+    auto callee = Apply(expr).callee();
+    return callee.kind() == TK_VAR && Var(callee).name().name() == "isinstance";
+  }
+
+  bool isPotentialNoneCheck(const Expr& expr) {
+    return expr.kind() == TK_IS || expr.kind() == TK_ISNOT;
+  }
+
   void emitIf(const If& stmt) {
-    // NOTE: emitIf checks on If stmt condition to see if the cond AST kind ==
-    // is/is not, for such cases we do meta programming and disable emitting the
+    // NOTE: emitIf checks on If stmt condition to see if the cond AST is
+    // a potential none check with is/is not, or an isinstance check.
+    // for such cases we do meta programming and disable emitting the
     // corresponding branches
     Expr cond = stmt.cond();
+    bool isinstance_call = isIsInstanceCall(cond);
+    bool potential_none_check = !isinstance_call && isPotentialNoneCheck(cond);
```

Comment:
nit: Is `isIsInstanceCall(cond)` ever true when `isPotentialNoneCheck(cond)` is true? Otherwise, seems like this conditional can just be `bool potential_none_check = isPotentialNoneCheck(cond)`.

Reply:
This is maybe over-optimization.  `isPotentialNoneCheck(cond)` is never true when `isIsInstanceCall(cond)` is true. So here I am only computing `isPotentialNoneCheck` when `isIsInstanceCall` is false.

If you think i should remove it i will